testgen/src/light_block.rs

```diff
@@ -0,0 +1,249 @@
+        self.header
+            .as_ref()
+            .expect("header is missing")
+            .chain_id
+            .as_ref()
+            .expect("chain_id is missing")
+            .to_string()
+    }
+}
+
+impl std::str::FromStr for LightBlock {
+    type Err = SimpleError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let testgen_light_block = match parse_as::<LightBlock>(s) {
+            Ok(input) => input,
+            Err(_) => LightBlock::new_default(parse_as::<Vec<Validator>>(s)?.as_ref(), 1),
+        };
+        Ok(testgen_light_block)
+    }
+}
+
+impl Generator<TMLightBlock> for LightBlock {
+    fn merge_with_default(self, default: Self) -> Self {
+        Self {
+            header: self.header.or(default.header),
+            commit: self.commit.or(default.commit),
+            validators: self.validators.or(default.validators),
+            next_validators: self.next_validators.or(default.next_validators),
+            provider: self.provider.or(default.provider),
+        }
+    }
+
+    fn generate(&self) -> Result<TMLightBlock, SimpleError> {
+        let header = match &self.header {
+            None => bail!("header is missing"),
+            Some(h) => h,
+        };
+        let commit = match &self.commit {
+            None => bail!("commit is missing"),
+            Some(c) => c,
+        };
+        let signed_header =
+            generate_signed_header(header, commit).expect("Could not generate signed header");
+
+        let validators = match &self.validators {
+            None => bail!("validator array is missing"),
+            Some(vals) => validator::Set::new(generate_validators(vals)?),
+        };
+
+        let next_validators = match &self.next_validators {
+            Some(next_vals) => validator::Set::new(generate_validators(next_vals)?),
+            None => validators.clone(),
+        };
+
+        let provider = match PeerId::from_str(
+            self.provider
+                .clone()
+                .unwrap_or_else(|| "peer-1".to_string())
+                .as_str(),
+        ) {
+            Ok(id) => id,
+            Err(_) => bail!("failed to construct light block's peer_id"),
+        };
+
+        let light_block = TMLightBlock {
+            signed_header,
+            validators,
+            next_validators,
+            provider,
+        };
+
+        Ok(light_block)
+    }
+}
+
+/// A helper function to generate SignedHeader used by TMLightBlock
+pub fn generate_signed_header(
+    raw_header: &Header,
+    raw_commit: &Commit,
+) -> Result<SignedHeader, SimpleError> {
+    let header = match raw_header.generate() {
+        Err(e) => bail!("Failed to generate header with error: {}", e),
+        Ok(h) => h,
+    };
+
+    let commit = match raw_commit.generate() {
+        Err(e) => bail!("Failed to generate commit with error: {}", e),
+        Ok(c) => c,
+    };
+
+    Ok(SignedHeader { header, commit })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```

Comment:
Could we also have a test that describes the conversion from a `LightBlock` into a `TMLightBlock`? It can also include some calls to `next` and `new_default` like in the test below.

Reply:
done ✅